archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,872 @@
+/-
+Copyright (c) 2022. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching, Fabian Kruse, Nikolas Kuhn
+-/
+import analysis.special_functions.log.basic
+import analysis.special_functions.pow
+import algebra.big_operators.basic
+import algebra.big_operators.intervals
+import data.finset.sum
+import data.fintype.basic
+import data.real.basic
+import data.real.pi.wallis
+import order.filter
+import order.filter.basic
+import order.bounded_order
+import topology.instances.real
+import topology.instances.ennreal
+
+/-!
+# Stirling's formula
+
+This file proves Theorem 90 from the [100 Theorem List] <https://www.cs.ru.nl/~freek/100/>.
+It states that $n!$ grows asymptotically like $\sqrt{2\pi n}(\frac{n}{e})^n$.
+
+
+## Proof outline
+
+The proof follows: <https://proofwiki.org/wiki/Stirling%27s_Formula>.
+
+We proceed in two parts.
+### Part 1
+We consider the fraction sequence $a_n$ of fractions $n!$ over $\sqrt{2n}(\frac{n}{e})^n$ and
+proves that this sequence converges against a real, positve number $a$. For this the two main
+ingredients are
+ - taking the logarithm of the sequence and
+ - use the series expansion of $\log(1 + x)$.
+
+### Part 2
+We use the fact that the series defined in part 1 converges againt a real number $a$ and prove that
+$a = \sqrt{\pi}$. Here the main ingredient is the convergence of the Wallis product.
+-/
+
+
+open_locale big_operators -- notation ∑ for finite sums
+open_locale classical real topological_space nnreal ennreal filter big_operators
+open  finset filter nat real
+
+namespace stirling
+
+/-- TODO: Perhaps something to add as rat.cast_sum in more generality (see below) in mathlib?!-/
+lemma rat_cast_sum (s : finset ℕ) (f : ℕ → ℚ) :
+  ↑(∑ n in s, f n : ℚ) = (∑ n in s, (f n : ℝ)) :=
+  (rat.cast_hom ℝ).map_sum f s
+-- @[simp, norm_cast] lemma rat_cast_sum [add_comm_monoid β] [has_one β]
+-- (s : finset α) (f : α → ℚ) :
+-- ↑(∑ x in s, f x : ℚ) = (∑ x in s, (f x : β)) := by sorry
+
+/-- **Sum of the Reciprocals of the Triangular Numbers**
+ (copied from archive)
+ TODO: include in some form mathlib or figure out how to import it from archive/ --/
+lemma inverse_triangle_sum :
+  ∀ n, ∑ k in range n, (2 : ℚ) / (k * (k + 1)) = if n = 0 then 0 else 2 - (2 : ℚ) / n :=
+begin
+  refine sum_range_induction _ _ (if_pos rfl) _,
+  rintro (_|n), { rw [if_neg, if_pos]; norm_num },
+  simp_rw [if_neg (succ_ne_zero _), succ_eq_add_one],
+  have A : (n + 1 + 1 : ℚ) ≠ 0, by { norm_cast, norm_num },
+  push_cast,
+  field_simp [cast_add_one_ne_zero],
+  ring,
+end
+
+/-- **Sum of products of consecutive reciprocals** -/
+lemma partial_sum_consecutive_reciprocals :
+ ∀ n, ∑ k in range n, (1 : ℚ) / (k.succ * (k.succ.succ)) ≤ 1 :=
```

Comment:
These lemmas about triangular numbers are nice, but I think you could defer them to a separate project.  All you need for this project is the summability of say `1 / (k.succ)^2`, right?  (without a precise knowledge of what the sum is).  And this is already in the library as [real.summable_one_div_nat_rpow](https://leanprover-community.github.io/mathlib_docs/analysis/p_series.html#real.summable_one_div_nat_rpow).

Reply:
sorry closed, but it is still a bit clumsy...